### PR TITLE
store/tikv: refactor connection pool for grpc

### PR DESCRIPTION
**Pull request by hhkbp2**
Hi,

This PR refactors connection pool for grpc. It takes advantage of that the grpc implementation supports multiple requests in one connection. And it removes the limitation of resource pool size for the capacity of concurrent requests.

PTAL @tiancaiamao @disksing @siddontang @ngaut

### CODE REVIEW

File: store/tikv/conn_map.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+// limitations under the License.
+
+package tikv
+
+import (
+	"fmt"
+	"sync"
+
+	"google.golang.org/grpc"
+
```
Comment (tiancaiamao) [inline store/tikv/conn_map.go]: remove empty line

File: store/tikv/conn_map.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	"github.com/juju/errors"
+)
+
+// Conn is a simple wrapper of grpc.ClientConn.
+type Conn struct {
+	c *grpc.ClientConn
+}
+
+// createConnFunc is the type of functions that can be used to create a grpc.ClientConn.
+type createConnFunc func(addr string) (*grpc.ClientConn, error)
```
Comment (tiancaiamao) [inline store/tikv/conn_map.go]: No need to define this, we can `grpc.Dial`

File: store/tikv/conn_map.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+		conn = &Conn{
+			c: c,
+		}
+		p.m.conns[addr] = conn
+	}
+	return conn, nil
+}
+
+// Put puts a Conn back to the map by the specific addr.
+func (p *ConnMap) Put(addr string, c *grpc.ClientConn) {
```
Comment (tiancaiamao) [inline store/tikv/conn_map.go]: Put is not need any more!

**Comment by disksing** [comment]
I think there are still too many abstractions. IMO, just change _rpcClient_ to
```
struct rpcClient {
    sync.RWMutex
    conns map[string]*ClientConn
}
```
then add a _func getConn(addr string) *ClientConn_ to it should be enough.

**Comment by hhkbp2** [comment]
PTAL @tiancaiamao @disksing @siddontang @ngaut

File: store/tikv/client.go [hunk 4]
```diff
@@ -68,24 +121,22 @@ func (c *rpcClient) SendReq(ctx goctx.Context, addr string, req *tikvrpc.Request
...
 
-	conn, err := c.p.GetConn(addr)
+	conn, err := c.getConn(addr)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	connPoolHistogram.WithLabelValues("cop").Observe(time.Since(start).Seconds())
-	defer c.p.PutConn(conn)
-	resp, err := c.callRPC(ctx, conn, req)
+	connMapHistogram.WithLabelValues(label).Observe(time.Since(start).Seconds())
```
Comment (disksing) [inline store/tikv/client.go]: Can we remove this metrics? It was originally used for the conn pool.

**Comment by disksing** [comment]
LGTM. PTAL @tiancaiamao

**Comment by tiancaiamao** [comment]
LGTM

File: store/tikv/client.go [hunk 5]
```diff
@@ -46,19 +48,70 @@ type Client interface {
...
+	c.Lock()
+	defer c.Unlock()
+	conn, ok := c.conns[addr]
+	if !ok {
+		var err error
+		conn, err = grpc.Dial(
+			addr,
+			grpc.WithInsecure(),
+			grpc.WithTimeout(dialTimeout),
+			grpc.WithUnaryInterceptor(grpc_prometheus.UnaryClientInterceptor),
```
Comment (siddontang) [inline store/tikv/client.go]: seem we can remove some customized metrics now,  @disksing

**Pull request merged**